File: src/deadlock_brain/sources/wiki.py
from __future__ import annotations

import json
import time
import urllib.parse
from pathlib import Path
from typing import Any

from deadlock_brain.http import HttpClient
from deadlock_brain.storage import BrainStore


SOURCE = "deadlock_wiki"
DEFAULT_API_URL = "https://deadlock.wiki/api.php"


class WikiRateLimiter:
    def __init__(self, state_path: Path, *, min_delay_seconds: float) -> None:
        self.state_path = state_path
        self.min_delay_seconds = min_delay_seconds
        self.state_path.parent.mkdir(parents=True, exist_ok=True)

    def wait(self) -> None:
        last = 0.0
        if self.state_path.exists():
            try:
                last = float(self.state_path.read_text(encoding="utf-8").strip() or "0")
            except ValueError:
                last = 0.0
        wait_for = self.min_delay_seconds - (time.time() - last)
        if wait_for > 0:
            time.sleep(wait_for)
        self.state_path.write_text(str(time.time()), encoding="utf-8")


def pull_wiki_page(
    store: BrainStore,
    http: HttpClient,
    *,
    title: str,
    enabled: bool,
    cache_ttl_seconds: int,
    min_delay_seconds: float,
    api_url: str = DEFAULT_API_URL,
) -> dict[str, Any]:
    if not enabled:
        raise RuntimeError(
            "Wiki-Netzwerkzugriff ist deaktiviert. Setze DEADLOCK_BRAIN_WIKI_ENABLED=1 "
            "oder nutze die CLI-Option --allow-wiki-network."
        )

    limiter = WikiRateLimiter(http.cache_dir / "wiki_last_request.txt", min_delay_seconds=min_delay_seconds)
    limiter.wait()
    params = urllib.parse.urlencode(
        {
            "action": "query",
            "format": "json",
            "prop": "extracts|revisions",
            "explaintext": "1",
            "rvprop": "ids|timestamp|content",
            "rvslots": "main",
            "titles": title,
            "redirects": "1",
        }
    )
    url = f"{api_url}?{params}"
    result = http.get(url, cache_ttl_seconds=cache_ttl_seconds, timeout=30)
    payload = json.loads(result.text)
    raw = json.dumps(payload, ensure_ascii=True, sort_keys=True).encode("utf-8")
    raw_path = store.write_raw(source=SOURCE, external_id=title, content=raw, suffix="json")
    document_id = store.upsert_source_document(
        source=SOURCE,
        external_id=title,
        title=title,
        url=url,
        content_type="application/json",
        raw_path=raw_path,
        content=raw,
        metadata={"from_cache": result.from_cache, "cache_ttl_seconds": cache_ttl_seconds},
    )
    store.upsert_entity_snapshot(
        source=SOURCE,
        entity_type="wiki_page",
        external_id=title,
        canonical_name=title,
        payload=payload,
        source_document_id=document_id,
    )
    pages = payload.get("query", {}).get("pages", {})
    return {
        "title": title,
        "from_cache": result.from_cache,
        "pages": len(pages) if isinstance(pages, dict) else 0,
    }

File: rust/crates/dbrain-sources/src/wiki.rs
use std::{
    fs,
    path::{Path, PathBuf},
    thread,
    time::{Duration, SystemTime, UNIX_EPOCH},
};

use deadlock_brain_core::http::{HttpClient, HttpGetOptions};
use serde_json::{json, Value};

use crate::{
    store::{complete_run, json_bytes, open_pool, EntitySnapshotInput, SourceDocumentInput, SourceStore},
    util::form_urlencode,
    Result, SourcesError,
};

pub const SOURCE: &str = "deadlock_wiki";
pub const DEFAULT_API_URL: &str = "https://deadlock.wiki/api.php";

#[derive(Debug, Clone)]
pub struct PullWikiPageOptions {
    pub title: String,
    pub enabled: bool,
    pub cache_ttl_seconds: u64,
    pub min_delay_seconds: f64,
    pub api_url: String,
}

impl Default for PullWikiPageOptions {
    fn default() -> Self {
        Self {
            title: String::new(),
            enabled: false,
            cache_ttl_seconds: 604_800,
            min_delay_seconds: 5.0,
            api_url: DEFAULT_API_URL.to_string(),
        }
    }
}

pub async fn pull_wiki_page(
    raw_dir: &Path,
    http: &HttpClient,
    options: PullWikiPageOptions,
) -> Result<Value> {
    let pool = open_pool().await?;
    let store = SourceStore::new(&pool, raw_dir)?;
    let run_id = store.begin_run("wiki").await?;
    let outcome = pull_wiki_page_inner(&store, http, &options).await;
    complete_run(&store, run_id, outcome).await
}

async fn pull_wiki_page_inner(
    store: &SourceStore<'_>,
    http: &HttpClient,
    options: &PullWikiPageOptions,
) -> Result<Value> {
    if !options.enabled {
        return Err(SourcesError::invalid_input(
            "Wiki-Netzwerkzugriff ist deaktiviert. Setze DEADLOCK_BRAIN_WIKI_ENABLED=1 \
oder nutze die CLI-Option --allow-wiki-network.",
        ));
    }

    let limiter = WikiRateLimiter {
        state_path: http.cache_dir().join("wiki_last_request.txt"),
        min_delay_seconds: options.min_delay_seconds,
    };
    limiter.wait()?;
    let params = [
        ("action", "query".to_string()),
        ("format", "json".to_string()),
        ("prop", "extracts|revisions".to_string()),
        ("explaintext", "1".to_string()),
        ("rvprop", "ids|timestamp|content".to_string()),
        ("rvslots", "main".to_string()),
        ("titles", options.title.clone()),
        ("redirects", "1".to_string()),
    ];
    let url = format!("{}?{}", options.api_url, form_urlencode(&params));
    let result = http.get(
        &url,
        HttpGetOptions {
            cache_ttl_seconds: Some(options.cache_ttl_seconds),
            timeout: Duration::from_secs(30),
            ..HttpGetOptions::default()
        },
    )?;
    let payload: Value = serde_json::from_str(&result.text())?;
    let raw = json_bytes(&payload)?;
    let raw_path = store.write_raw(SOURCE, &options.title, &raw, "json")?;
    let metadata = json!({
        "from_cache": result.from_cache,
        "cache_ttl_seconds": options.cache_ttl_seconds,
    });
    let document_id = store
        .upsert_source_document(SourceDocumentInput {
            source: SOURCE,
            external_id: &options.title,
            title: Some(&options.title),
            url: Some(&url),
            content_type: "application/json",
            raw_path: &raw_path,
            content: &raw,
            metadata: &metadata,
        })
        .await?;
    store
        .upsert_entity_snapshot(
            &EntitySnapshotInput {
                source: SOURCE.to_string(),
                entity_type: "wiki_page".to_string(),
                external_id: options.title.clone(),
                canonical_name: Some(options.title.clone()),
                payload: payload.clone(),
            },
            Some(document_id),
        )
        .await?;
    let pages = payload
        .get("query")
        .and_then(|query| query.get("pages"))
        .and_then(Value::as_object)
        .map(|pages| pages.len())
        .unwrap_or(0);
    Ok(json!({
        "title": options.title.clone(),
        "from_cache": result.from_cache,
        "pages": pages,
    }))
}

#[derive(Debug, Clone)]
struct WikiRateLimiter {
    state_path: PathBuf,
    min_delay_seconds: f64,
}

impl WikiRateLimiter {
    fn wait(&self) -> Result<()> {
        if let Some(parent) = self.state_path.parent() {
            fs::create_dir_all(parent)?;
        }
        let last = fs::read_to_string(&self.state_path)
            .ok()
            .and_then(|value| value.trim().parse::<f64>().ok())
            .unwrap_or(0.0);
        let now = now_seconds_f64()?;
        let wait_for = self.min_delay_seconds - (now - last);
        if wait_for.is_finite() && wait_for > 0.0 {
            thread::sleep(Duration::from_secs_f64(wait_for));
        }
        fs::write(&self.state_path, now_seconds_f64()?.to_string())?;
        Ok(())
    }
}

fn now_seconds_f64() -> Result<f64> {
    Ok(SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs_f64())
}
